runtime/router/trie.go

```diff
@@ -0,0 +1,293 @@
+}
+
+// Param is a url parameter where key is the url segment pattern (without :) and
+// value is the actual segment of a matched url.
+// e.g. url /foo/123 matches /foo/:id, the url param has key "id" and value "123"
+type Param struct {
+	Key, Value string
+}
+
+// Trie is a radix trie to store string value at given url path,
+// a trie node corresponds to an arbitrary path substring.
+type Trie struct {
+	root *tnode
+}
+
+type tnode struct {
+	key      string
+	value    http.Handler
+	children []*tnode
+}
+
+// NewTrie creates a new trie.
+func NewTrie() *Trie {
+	return &Trie{
+		root: &tnode{
+			key: "",
+		},
+	}
+}
+
+// Set sets the value for given path, returns error if path already set
+func (t *Trie) Set(path string, value http.Handler) error {
+	if path == "" || strings.Contains(path, "//") {
+		return errPath
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	// ignore trailing slash
+	path = strings.TrimSuffix(path, "/")
+
+	// validate "*"
+	if strings.Contains(path, "*") && !strings.HasSuffix(path, "/*") {
+		return errors.New("/* must be the last path segment")
+	}
+	if strings.Count(path, "*") > 1 {
+		return errors.New("path can not contain more than one *")
+	}
+
+	err := t.root.set(path, value)
+	if e, ok := err.(*paramMismatch); ok {
+		return fmt.Errorf("path %q has a different param key %q, it should be the same key %q as in existing path %q", path, e.actual, e.expected, e.existingPath)
+	}
+	return err
+}
+
+// Get returns the value for given path, returns error if not found
+func (t *Trie) Get(path string) (http.Handler, []Param, error) {
+	if path == "" || strings.Contains(path, "//") {
+		return nil, nil, errPath
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	// ignore trailing slash
+	path = strings.TrimSuffix(path, "/")
+	return t.root.get(path)
+}
+
+func (t *tnode) set(path string, value http.Handler) error {
+	// find the longest common prefix
+	var l, i int
+	m, n := len(t.key), len(path)
+	if m > n {
+		l = n
+	} else {
+		l = m
+	}
+	for i < l && t.key[i] == path[i] {
+		i++
+	}
+
+	// find index j, k in key and path to which they match,
+	// j and k is only useful to check if there is conflict,
+	// they are not where splits happen, splits happen at index i.
+	var j, k int
+	for j < m && k < n {
+		if t.key[j] == ':' || path[k] == ':' {
+			oj, ok := j, k
+			same := t.key[j] == path[k]
+			for j < m && t.key[j] != '/' {
+				j++
+			}
+			for k < n && path[k] != '/' {
+				k++
+			}
+			if same && (j-oj) != (k-ok) {
+				return &paramMismatch{
+					t.key[oj:j],
+					path[ok:k],
+					t.key,
+				}
+			}
+		} else if t.key[j] == path[k] {
+			j++
+			k++
+		} else {
+			break
+		}
+	}
+
+	// conflicts caused by ":" is only possible when j == m
+	if j == m {
+		for _, c := range t.children {
+			if c.key[0] == path[k] || c.key[0] == ':' || path[k] == ':' {
+				if _, _, err := c.get(path[k:]); err == nil {
+					return errExist
```

Comment:
You will probably want to print out the match that conflicted for diagnostic purposes.